storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -88,85 +488,214 @@ type ReconcileVolume struct {
+		diskSize, err := self.getDiskSizeForVolume(ctx, volume)
 		if err != nil {
-			return reconcile.Result{}, err
+			reqLogger.Error(err, "call to GetVolumeSize failed")
+			return self.setFailedVolumeStatus(
+				ctx, volume, storagev1alpha1.CreationError,
+				"call to GetVolumeSize failed: %s", err.Error(),
+			)
+		}
+		pv := newPersistentVolume(volume, diskSize)
+		if err := self.createPersistentVolume(ctx, pv, volume); err != nil {
+			return delayedRequeue(err)
 		}
+		reqLogger.Info(
+			"creating a new PersistentVolume", "PersistentVolume.Name", pv.Name,
+		)
+		self.recorder.Event(
+			volume, corev1.EventTypeNormal, "PvCreation",
+			"backing PersistentVolume created",
+		)
+		return self.setAvailableVolumeStatus(ctx, volume)
+	default: // PrepareVolume in progress: poll its state.
+		return self.pollSaltJob(
+			ctx, "PrepareVolume", volume, self.setPendingVolumeStatus,
+		)
+	}
+}
 
-		// PV created successfully - don't requeue
-		return reconcile.Result{}, nil
-	} else if err != nil {
-		return reconcile.Result{}, err
+// Finalize a volume marked for deletion.
+func (self *ReconcileVolume) finalizeVolume(
+	ctx context.Context,
+	request reconcile.Request,
+	volume *storagev1alpha1.Volume,
+) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Name", request.Name)
+
+	if volume.Status.Phase == storagev1alpha1.VolumePending {
+		reqLogger.Info("pending volume cannot be marked for deletion: requeue")
+		return delayedRequeue(nil)
 	}
 
-	// PV already exists - don't requeue
-	reqLogger.Info("Skip reconcile: PersistentVolume already exists", "PersistentVolume.Name", found.Name)
+	// Check if a PV is associated to the volume.
+	pv, err := self.getPersistentVolume(ctx, volume.Name)
+	if err != nil {
+		reqLogger.Error(
+			err, "cannot read PersistentVolume: requeue",
+			"PersistentVolume.Name", volume.Name,
+		)
+		return delayedRequeue(err)
+	}
 
-	return reconcile.Result{}, nil
+	// No associated PV: let's delete ourselves then.
+	if pv == nil {
+		reqLogger.Info("no PersistentVolume associated to terminating volume")
+		if err := self.removeVolumeFinalizer(ctx, volume); err != nil {
+			reqLogger.Error(err, "cannot remove volume finalizer")
+			return delayedRequeue(err)
+		}
+		reqLogger.Info("volume finalizer removed")
+		return endReconciliation()
+	}
+
+	switch pv.Status.Phase {
+	case corev1.VolumeBound:
+		reqLogger.Info(
+			"backing PersistentVolume is bound: cannot delete volume",
+		)
+		return delayedRequeue(nil)
+	case corev1.VolumePending:
+		reqLogger.Info(
+			"backing PersistentVolume is pending: waiting for stabilization",
+		)
+		return delayedRequeue(nil)
+	case corev1.VolumeAvailable, corev1.VolumeReleased, corev1.VolumeFailed:
+		reqLogger.Info("the backing PersistentVolume is removable")
+		return self.destroyPersistentVolume(ctx, request, volume, pv)
+	default:
+		reqLogger.Info(
+			"unexpected backing PersistentVolume status: do nothing",
```

Comment:
Can we stringify the status into this log message?

Reply:
Done